File: jem-core/src/main/java/jem/util/text/Texts.java
/*
 * Copyright 2014-2017 Peng Wan <phylame@163.com>
 *
 * This file is part of Jem.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jem.util.text;

import jem.util.Variants;
import jem.util.flob.Flob;
import lombok.NonNull;
import lombok.SneakyThrows;
import lombok.val;
import pw.phylame.commons.io.IOUtils;
import pw.phylame.commons.util.Exceptions;

import java.io.IOException;
import java.io.Writer;
import java.nio.charset.Charset;
import java.util.Iterator;
import java.util.List;

import static pw.phylame.commons.util.StringUtils.EMPTY_TEXT;

/**
 * Factory class for creating {@code Text} instance.
 */
public final class Texts {

    private Texts() {
    }

    /**
     * Type for html text.
     */
    public static final String HTML = "html";

    /**
     * Type for plain text.
     */
    public static final String PLAIN = "plain";

    public static Text forEmpty(String type) {
        return forString(EMPTY_TEXT, type);
    }

    public static Text forString(@NonNull CharSequence cs, String type) {
        return new RawText(cs, type);
    }

    public static Text forFlob(@NonNull Flob file, String encoding, String type) {
        return new FlobText(file, encoding, type);
    }

    private static class RawText extends AbstractText {
        static {
            Variants.mapType(RawText.class, Variants.TEXT);
        }

        private final CharSequence text;

        private RawText(CharSequence cs, String type) {
            super(type);
            this.text = cs;
        }

        @Override
        public String getText() {
            return text.toString();
        }
    }

    private static class FlobText extends AbstractText {
        static {
            Variants.mapType(FlobText.class, Variants.TEXT);
        }

        private final Flob file;
        private final String encoding;

        private FlobText(Flob file, String encoding, String type) {
            super(type);
            this.file = file;
            this.encoding = encoding;
            if (encoding != null && !Charset.isSupported(encoding)) {
                throw Exceptions.forIllegalArgument("charset %s is unsupported", encoding);
            }
        }

        @Override
        @SneakyThrows(IOException.class)
        public String getText() {
            try (val in = file.openStream()) {
                return IOUtils.toString(in, encoding);
            }
        }

        @Override
        @SneakyThrows(IOException.class)
        public List<String> getLines(boolean skipEmpty) {
            try (val in = file.openStream()) {
                return IOUtils.toLines(in, encoding, skipEmpty);
            }
        }

        @Override
        @SneakyThrows(IOException.class)
        public Iterator<String> iterator() {
            try (val in = file.openStream()) {
                return IOUtils.linesOf(in, encoding, false);
            }
        }

        @Override
        public long writeTo(@NonNull Writer writer) throws IOException {
            try (val reader = IOUtils.readerFor(file.openStream(), encoding)) {
                return IOUtils.copy(reader, writer, -1);
            }
        }
    }
}

File: commons/src/main/kotlin/jclp/value/Texts.kt
package jclp.value

import java.io.Writer
import java.nio.charset.Charset

const val TEXT_HTML = "html"
const val TEXT_PLAIN = "plain"

interface Text : Iterable<String> {
    val type get() = TEXT_PLAIN

    override fun toString(): String

    override fun iterator() = toString().split("\r\n", "\n", "\r").iterator()

    fun writeTo(output: Writer) = toString().let {
        output.write(it)
        it.length.toLong()
    }

    companion object {
        fun empty(type: String = TEXT_PLAIN) = of("", type)

        fun of(cs: CharSequence, type: String = TEXT_PLAIN): Text = RawText(type, cs)

        fun of(flob: Flob, encoding: String? = null, type: String = TEXT_PLAIN): Text = FlobText(type, flob, encoding)
    }
}

open class TextWrapper(private val text: Text) : Text {
    override val type = text.type

    override fun toString() = text.toString()

    override fun iterator() = text.iterator()

    override fun writeTo(output: Writer) = text.writeTo(output)

    override fun equals(other: Any?) = text == other

    override fun hashCode() = text.hashCode()
}

abstract class AbstractText(final override val type: String) : Text {
    init {
        require(type.isNotEmpty()) { "type cannot be empty" }
    }
}

private class RawText(type: String, private val text: CharSequence) : AbstractText(type) {
    override fun toString() = text.toString()
}

private class FlobText(type: String, private val flob: Flob, encoding: String? = null) : AbstractText(type) {
    private val charset: Charset = when {
        encoding == null -> Charset.defaultCharset()
        encoding.isNotEmpty() -> Charset.forName(encoding)
        else -> throw IllegalArgumentException("Unsupported encoding: $encoding")
    }

    override fun toString() = openReader().readText()

    override fun iterator() = openReader().readLines().iterator()

    override fun writeTo(output: Writer) = openReader().copyTo(output)

    private fun openReader() = flob.openStream().bufferedReader(charset)
}
